File: dbassistant/tokens/Tokenizer.py
from dbassistant.interfaces.IGlossaryHandlerFactory import IGlossaryHandlerFactory
from dbassistant.tokenhandlers.ArgumentTokenHandler import ArgumentTokenHandler
from dbassistant.tokenhandlers.EmptyTokenHandler import EmptyTokenHandler
from dbassistant.tokenhandlers.EndTokenHandler import EndTokenHandler
from dbassistant.tokenhandlers.GenericTokenHandler import GenericTokenHandler
from dbassistant.tokens.TokenSeries import TokenSeries
from dbassistant.tokens.TokenType import TokenType


class Tokenizer:
    def __init__(
        self,
        glossaryHandlerFactory: IGlossaryHandlerFactory,
        promptSize: int,
        maxArgumentCount: int,
        maxGenericTokenCount: int
    ):
        self.glossaryHandlerFactory = glossaryHandlerFactory
        self.promptSize = promptSize
        self.maxArgumentCount = maxArgumentCount
        self.maxGenericTokenCount = maxGenericTokenCount
        
        self.endTokenHandler = EndTokenHandler(0)
        self.emptyTokenHandler = EmptyTokenHandler(1)
        self.glossaryTokenHandler = glossaryHandlerFactory.build(2)
        self.genericTokenHandler = GenericTokenHandler(3, maxGenericTokenCount)
        self.argumentTokenHandler = ArgumentTokenHandler(10, maxArgumentCount)
        
        self.handlers = [
            self.endTokenHandler,
            self.emptyTokenHandler,
            self.argumentTokenHandler,
            self.genericTokenHandler,
            self.glossaryTokenHandler
        ]
        
    @property
    def maxSize(self):
        return sum(h.size for h in self.handlers)
    
    @property
    def utilizedSize(self):
        return sum(h.utilizedSize for h in self.handlers)
    
    def encode(self, input):
        tokens = []
        
        for index in range(len(input)):
            word = input[index]
            
            for handler in sorted(self.handlers, key=lambda h: h.priority):
                if handler.canEncode(word):
                    offset = self.handlerOffset(handler)
                    token = handler.encode(word)
                    tokens.append(offset + token.value)
                    
                    if token.type == TokenType.KEY:
                        argOffset = self.handlerOffset(self.argumentTokenHandler)
                        argToken = self.argumentTokenHandler.encode(word)
                        tokens.append(argOffset + argToken.value)
                    
                    break
        
        # Possible exception: generated tokens are longer than the input words
        for index in range(len(tokens), self.promptSize):
            tokens.append(self.handlerOffset(self.emptyTokenHandler))
        
        return TokenSeries(tokens)
    
    def decode(self, input):
        words = []
        
        for token in input.tokens:
            offset = 0
            for handler in self.handlers:
                if token < offset + handler.size:
                    word = handler.decode(token - offset)
                    words.append(word)
                    break
                else:
                    offset += handler.size
        
        return words
    
    def resetState(self):
        for handler in self.handlers:
            handler.resetState()
    
    def serialize(self, path):
        for handler in self.handlers:
            handler.serialize(path)
    
    def deserialize(self,path):
        for handler in self.handlers:
            handler.deserialize(path)
        
    
    def handlerOffset(self, handler):
        return sum(h.size for h in self.handlers[:self.handlers.index(handler)])
    
    def copyUntrained(self):
        return Tokenizer(
            self.glossaryHandlerFactory,
            self.promptSize,
            self.maxArgumentCount,
            self.maxGenericTokenCount
        )

File: dbassistant/tokens/Tokenizer.ts
import IGlossaryHandlerFactory from '../interfaces/IGlossaryHandlerFactory';
import ITokenHandler from '../interfaces/ITokenHandler';
import {ArgumentTokenHandler} from '../tokenhandlers/ArgumentTokenHandler';
import {EmptyTokenHandler} from '../tokenhandlers/EmptyTokenHandler';
import {EndTokenHandler} from '../tokenhandlers/EndTokenHandler';
import {GenericTokenHandler} from '../tokenhandlers/GenericTokenHandler';
import {GlossaryTokenHandler} from '../tokenhandlers/GlossaryTokenHandler';
import TokenSeries from '../tokens/TokenSeries';
import TokenType from '../tokens/TokenType';

class Tokenizer {
    private glossaryHandlerFactory: IGlossaryHandlerFactory;
    private promptSize: number;
    private maxArgumentCount: number;
    private maxGenericTokenCount: number;
    private endTokenHandler: EndTokenHandler;
    private emptyTokenHandler: EmptyTokenHandler;
    public glossaryTokenHandler: GlossaryTokenHandler;
    private genericTokenHandler: GenericTokenHandler;
    private argumentTokenHandler: ArgumentTokenHandler;
    private handlers: ITokenHandler[];

    constructor(
        glossaryHandlerFactory: IGlossaryHandlerFactory,
        promptSize: number,
        maxArgumentCount: number,
        maxGenericTokenCount: number
    ) {
        this.glossaryHandlerFactory = glossaryHandlerFactory;
        this.promptSize = promptSize;
        this.maxArgumentCount = maxArgumentCount;
        this.maxGenericTokenCount = maxGenericTokenCount;

        //encode order by priority
        this.endTokenHandler = new EndTokenHandler(0);
        this.emptyTokenHandler = new EmptyTokenHandler(1);
        this.glossaryTokenHandler = glossaryHandlerFactory.build(2);
        this.genericTokenHandler = new GenericTokenHandler(3, maxGenericTokenCount);
        this.argumentTokenHandler = new ArgumentTokenHandler(10, maxArgumentCount);

        this.handlers = [
            // decode order
            this.endTokenHandler,
            this.emptyTokenHandler,
            this.argumentTokenHandler,
            this.genericTokenHandler,
            this.glossaryTokenHandler,
        ];
    }

    get maxSize(): number {
        return this.handlers.reduce((acc, h) => acc + h.size, 0);
    }

    get labelCount(): number {
        return this.maxSize - this.glossaryTokenHandler.keyCount;
    }

    get utilizedSize(): number {
        return this.handlers.reduce((acc, h) => acc + h.utilizedSize, 0);
    }

    encode(input: string[]): TokenSeries {
        const tokens: number[] = [];

        for (let index = 0; index < input.length; index++) {
            const word = input[index];

            for (const handler of this.handlers.sort((a, b) => a.priority - b.priority)) {
                if (handler.canEncode(word)) {
                    const offset = this.handlerOffset(handler);
                    const token = handler.encode(word);
                    tokens.push(offset + token.value);

                    if (token.type === TokenType.KEY) {
                        const argOffset = this.handlerOffset(this.argumentTokenHandler);
                        const argToken = this.argumentTokenHandler.encode(word);
                        tokens.push(argOffset + argToken.value);
                    }

                    break;
                }
            }
        }

        // Possible exception: generated tokens are longer than the input words
        for (let index = tokens.length; index < this.promptSize; index++) {
            tokens.push(this.handlerOffset(this.emptyTokenHandler));
        }

        return new TokenSeries(tokens);
    }

    decode(input: TokenSeries): string[] {
        const words: string[] = [];

        for (const token of input.tokens) {
            let offset = 0;
            for (const handler of this.handlers) {
                if (token < offset + handler.size) {
                    const word = handler.decode(token - offset);
                    words.push(word);
                    break;
                } else {
                    offset += handler.size;
                }
            }
        }

        return words;
    }

    resetState(): void {
        for (const handler of this.handlers) {
            handler.resetState();
        }
    }

    serialize(path: string): void {
        for (const handler of this.handlers) {
            const serialized = handler.serialize(path);
        }
    }

    deserialize(path: string): void {
        for (const handler of this.handlers) {
            handler.deserialize(path);
        }
    }

    private handlerOffset(handler: any): number {
        return this.handlers.slice(0, this.handlers.indexOf(handler)).reduce((acc, h) => acc + h.size, 0);
    }

    copyUntrained(): Tokenizer {
        return new Tokenizer(
            this.glossaryHandlerFactory,
            this.promptSize,
            this.maxArgumentCount,
            this.maxGenericTokenCount
        );
    }
}

export default Tokenizer;
